### ISSUE TITLE
Rengine not being shutdown correctly on reload of service

### ISSUE THREAD

**Issue by devilgate** [smallAreaHealthStatisticsUnit/rapidInquiryFacility#87]
Peter:

> Run a study, RIF service web application will fail to load R DLL as it is still attached to an old thread.

```
  Cannot find JRI native library!
  Please make sure that the JRI native library is in a directory listed in java.library.path.
  java.lang.UnsatisfiedLinkError: Native Library C:\Program Files\R\R-3.4.0\library\rJava\jri\x64\jri.dll already loaded in another classloader
        at java.lang.ClassLoader.loadLibrary0(Unknown Source)
        at java.lang.ClassLoader.loadLibrary(Unknown Source)
        at java.lang.Runtime.loadLibrary0(Unknown Source)
        at java.lang.System.loadLibrary(Unknown Source)
        at org.rosuda.JRI.Rengine.<clinit>(Rengine.java:19)
        at rifServices.dataStorageLayer.pg.PGSQLSmoothResultsSubmissionStep.performStep(PGSQLSmoothResultsSubmissionStep.java:183)
        at rifServices.dataStorageLayer.pg.PGSQLRunStudyThread.smoothResults(PGSQLRunStudyThread.java:257)
        at rifServices.dataStorageLayer.pg.PGSQLRunStudyThread.run(PGSQLRunStudyThread.java:176)
        at java.lang.Thread.run(Unknown Source)
        at rifServices.dataStorageLayer.pg.PGSQLAbstractRIFStudySubmissionService.submitStudy(PGSQLAbstractRIFStudySubmissionService
```
> The solution is to restart tomcat. Server reload needs to stop R. This requires a `@WebListener [Context Listener (javax.servlet.ServletContextListener)]`. There is a risk this will still not work as JRI itself may not detach the DLL and this will need to be done manually.

**Comment by devilgate** [comment]
It seems to me that we should be able to stop the R Engine running after each use. That might not be the most efficient way to do things, but it should prevent locking problems.

However, it's not obvious, from any examples I've found so far, how you should do that. There are various methods provided in various implementations of the engine, but I'm not sure how reliable they are.

Since the `REngine` class is a `Thread`, it might be best to use standard `Thread` techniques to make sure it stops.

I'm investigating further.

**Comment by devilgate** [comment]
I've now replicated this problem on Mac. The symptoms are not identical, but the sequence of events:

1. run a study;
2. deploy `rifServices.war` without stopping Tomcat first;
3. try to do log back on to the RIF;

results in a broken system. Restarting Tomcat puts everything back in order.

On the plus side, this should make it easy for me to test potential fixes.

**Comment by devilgate** [comment]
Having spent the day struggling with this, I'm coming to the conclusion that the problem is in JRI, or at least in our interaction with it. I've [opened an issue](https://github.com/s-u/rJava/issues/166) over there to see if we can get some help.

However, I think the only way we'll get round this is to implement the other approach, as suggested in #111: run the R functions in a separate service. At least if we do that we should be able to hot deploy `rifServices`. Of course, this problem will move over to the new service, but we'll just have to tell users that if they're deploying the R service they'll have to restart Tomcat.

**Comment by devilgate** [comment]
Closed by the merge of #114.
